### PR TITLE
[Install] Remove upper bound for NumPy for default installation and update upper bounds for extras

### PR DESCRIPTION
**Pull request by rkazants**
**Details:** In this PR we remove the upper-bound for NumPy in case of default installation (like `pip install openvino` and `pip install openvino-dev`) so that the user AI environment (for example, with NumPy and ML/DL frameworks) will not be affected and over-written.
In the meantime, we leave and update the upper-bounds for extra installation (like `pip install openvino-dev[tensorflow]`) that should provide installation along with the certain versions of AI third-party components against which `openvino` and `openvino-dev` were tested.

The following changes were done:
* remove upper-bound of NumPy for Python bindings, samples, and tools, in case the default installation
* update versions of NumPy, TensorFlow, NetworkX for extras and tests requirements

**Ticket:** 96506

Signed-off-by: Kazantsev, Roman <roman.kazantsev@intel.com>

### CODE REVIEW

**Comment by yury-gorbachev** [comment]
I do not understand why you keep introducing upper bound over and over - can you explain?

**Review by Wovchena — APPROVED** [review]

**Comment by rkazants** [comment]
> I do not understand why you keep introducing upper bound over and over - can you explain?

I am absolutely fine to remove the upper bound for some modules like NumPy and TensorFlow at all.
But in the previous PR I received concerns (from QA team, Python API stakeholders, NNCF/POT) about possible bad user experience in case forward incompatibility (future versions of these dependencies against OV). I heard about requirements for NNCF tool that strictly uses the concrete version of TF.

@AlexKoff88, can you comment on NNCF tool? Can we remove the upper bound for TF? It should not affect MO.
@jiwaszki, @ilya-lavrenov, propose to remove  NumPy upper bound version. Any concerns?

**Comment by ilyachur** [comment]
I also agree with removing upper bound

**Comment by rkazants** [comment]
@yury-gorbachev, removed upper bounds for TensorFlow and NumPy. For NetworkX I am leaning to update up to the latest stable version because I have no data how it affected us or not in the past during the update. Now there can be a problem with the validation that become sort of indeterministic in case new releases of TF and Numpy modules. We need to create a separate requirement files for the validation.

@akladiev, can you please comment on the changes in PR and can we make it deterministic by creating reqs files for the validation?

File: src/bindings/python/requirements.txt [hunk 1]
```diff
@@ -1 +1 @@
-numpy>=1.16.6,<=1.23.1
+numpy>=1.16.6
```
Comment (jiwaszki) [inline src/bindings/python/requirements.txt:1]:
```suggestion
numpy>=1.16.6,<=1.23.4
```
Please do it in all following `numpy` cases. Similar for Tensorflow.

We had discussed it together today.

**Review by jiwaszki — CHANGES_REQUESTED** [review]

File: src/bindings/python/requirements.txt (hunk 1 above)
Comment (rkazants) [reply src/bindings/python/requirements.txt:1]: Let us discuss offline the next steps. I would not create super story from this task.
Let us negotiate with QA team about the next steps

**Comment by AlexKoff88** [comment]
> > I do not understand why you keep introducing upper bound over and over - can you explain?
> 
> I am absolutely fine to remove the upper bound for some modules like NumPy and TensorFlow at all. But in the previous PR I received concerns (from QA team, Python API stakeholders, NNCF/POT) about possible bad user experience in case forward incompatibility (future versions of these dependencies against OV). I heard about requirements for NNCF tool that strictly uses the concrete version of TF.
> 
> @AlexKoff88, can you comment on NNCF tool? Can we remove the upper bound for TF? It should not affect MO. @jiwaszki, @ilya-lavrenov, propose to remove NumPy upper bound version. Any concerns?

We don't have the upper bound in NNCF. However, we have a recommended (validated) version that users can rely on when installing NNCF with extras. This change is ok from the NNCF standpoint.

File: src/bindings/python/requirements.txt (hunk 1 above)
Comment (rkazants) [reply src/bindings/python/requirements.txt]: FIxed.

**Comment by rkazants** [comment]
@yury-gorbachev, @AlexKoff88, @ilya-lavrenov, we decided not to remove the upper-bound for these dependencies this release due to some potential risks, obstacles from the validation.
Once we create a branch for 2022.3, I will immediately create PR to remove these upper-bound and get it merged to go without them the next release and to have the QA team prepared to these changes in advance.
Now we just update upper bounds.

Best regards,
Roman

**Comment by helena-intel** [comment]
Will we keep testing LTS releases regularly to ensure compatibility of OpenVINO Runtime with the very latest version of numpy? Two years is a long time, and we should make sure our LTS release does not break with a numpy upgrade. For OpenVINO Runtime, do we currently test both the lower and upper versions that are specified as compatible?

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by jiwaszki** [comment]
> When the upper bound is removed, will we keep testing LTS releases regularly to ensure compatibility of OpenVINO Runtime with the very latest version of numpy? Two years is a long time, and we should make sure our LTS release does not break with a numpy upgrade. For OpenVINO Runtime, do we currently test both the lower and upper versions that are specified as compatible?

This is one of major concerns. LTS releases will get a lot of fixes in periods like this. Especially considering major Python updates that are made or being planned in a span of 2 years. Thanks @helena-intel for bringing it up.

Additionally:

- `samples/python/requirements.txt` - `numpy` is unrestricted, do we ensure that samples are always working? What in the case of breakage? Hotfixing in additional releases? Sending messages to users to lower version? In my opinion, at least samples requirements should be restricted - ensuring their ability to work.
- `src/bindings/python/src/compatibility/openvino/requirements.txt` - I do not see a reason of compatibility bump here. However this is more of a issue of keeping both APIs coexisting. It was "promised" to be frozen, but at this point it is constantly updated.
- `tools/mo/requirements.txt` and `src/frontends/paddle/tests/requirements.txt` etc. - just to add to other comments. We are testing up to `numpy1.23.4` but allowing unrestricted package to users. There is already [a pre-release version of 1.24.0](https://pypi.org/project/numpy/1.24.0rc1/), is it going to be tested when final version is uploaded? As a user, I would expect it being tested and fully operational as requirements are unrestricted in common requirements.

I am not going to give it either green or red light as I see potential issues. Especially on tests side, where there is no clear information about range of tested packages versions.

PS is there a branch for 2022.3? If this PR is merged, it should be done **only after** the branch creation.

**Pull request merged**